compiler/rustc_span/src/def_id.rs

```diff
@@ -205,12 +203,19 @@ impl<D: Decoder> Decodable<D> for DefIndex {
 /// index and a def index.
 ///
 /// You can create a `DefId` from a `LocalDefId` using `local_def_id.to_def_id()`.
-#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Copy)]
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Copy)]
 pub struct DefId {
     pub krate: CrateNum,
     pub index: DefIndex,
 }
 
+// hash one u64 instead of two u32s
+impl Hash for DefId {
+    fn hash<H: Hasher>(&self, h: &mut H) {
+        ((u64::from(u32::from(self.krate)) << 32) | u64::from(u32::from(self.index))).hash(h)
```

Comment:
```suggestion
        ((u64::from(u32::from(self.index)) << 32) | u64::from(u32::from(self.krate))).hash(h)
```

It may be faster this way: https://rust.godbolt.org/z/EMeebaGb3

Reply:
Yes but that would exacerbate the problem of def index bit distribution further (at least my theory is that we usually have more defs in a crate than we have crates within one build, but I may be wrong).

I'd rather reorder the fields in the type.

Reply:
Yeah (probably 64-bit LE only to boot) -- if you throw a `--target` on the godbolt for a BE arch you can see what it looks like.

In practice the most popular architectures these days are LE (x86-64 and aarch64), but I guess if one were really motivated they could use `#[cfg]` to keep it optimal on both.

Reply:
Thanks! I always find endianness tricky to think about, just checking my reasoning :)